lib/Runtime/Base/ThreadContextTlsEntry.cpp

```diff
@@ -87,6 +87,11 @@ bool ThreadContextTLSEntry::TrySetThreadContext(ThreadContext * threadContext)
     if (entry == NULL)
     {
         Assert(!threadContext->IsThreadBound());
+#ifdef RECYCLER_WRITE_BARRIER
+#ifdef TARGET_64
+        Memory::RecyclerWriteBarrierManager::OnThreadInit();
```

Comment:
this will be called the second time in ThreadContext::EnsureRecycler. can we remove that one?

Reply:
I'm unsure about that. From what I can tell, `EnsureRecycler` will only do work ~once, the first time that it hasn't yet set up the Recycler. Prior to this change, that would essentially always coincide with the main thread which would have already called `OnThreadInit` from static initialization I think. I could remove the call from in there, but I'm not sure if there is some edge case I'm missing, and it also actually has some error handling around it, which this call and the `StaticBackgroundProc` instances of `OnThreadInit` don't. I would happily defer to someone who knows more about how this should work though.

Reply:
Why does this call not have error handling? I would argue that all calls to OnThreadInit should have error handling, and if that's the case, you can probably remove the call from EnsureRecycler.

Reply:
Currently there is a call in `BackgroundJobProcessor::StaticThreadProc` which doesn't have any error handling, and I'm not sure how you could handle errors there in a reasonable way.

I could change this call to `if (!Memory::RecyclerWriteBarrierManager::OnThreadInit()) { return false;}` and from the looks of things I think that should propagate correctly.